ibc_public/utils_labels.py

```diff
@@ -0,0 +1,162 @@
+import json
+import sys
+import warnings
+
+import numpy as np
+import pandas as pd
+
+import os
+
+_package_directory = os.path.dirname(os.path.abspath(__file__))
+
+ALL_CONTRASTS = os.path.join(_package_directory, '..', 'ibc_data',
+                             'all_contrasts.tsv')
+
+
+def get_labels(contrasts='all'):
+    """
+    Returns the list of labels for each passed contrast name
+
+    Parameters
+    ----------
+
+    contrasts: list of str, default 'all'
+               Each element of the list is a contrast name in the document.
+               The default argument will select all present contrasts
+
+    Returns
+    -------
+
+    contrast_dict: dict
+                   Dictionary containing the contrasts provided by the user
+                   as keys, and their corresponding labels as values
+    """
+    df = pd.read_csv(ALL_CONTRASTS, sep='\t')
+    if contrasts == 'all':
+        contrasts = df['contrast'].values
+
+    contrast_dict = {}
+
+    con_slice = df[df['contrast'].isin(contrasts)]
+    not_found = np.setdiff1d(contrasts, con_slice['contrast'])
+
+    if not_found.size != 0:
+        warnings.warn(f"The following contrast names were not found: "
+                      f"{not_found}")
+
+    for index, con in con_slice.iterrows():
+
+        labels = con.loc[con == 1.0].index
+        con_name = f"({con.task}) {con.contrast}"
+        contrast_dict[con_name] = [label for label in labels]
+
+    return contrast_dict
+
+
+def _binary_question(question):
+    """
+    Asks a yes/no question to the user and returns True or False depending
+    on the answer
+
+    Parameters
+    ----------
+
+    question: str
+              String prompted to the user
+
+    Returns
+    -------
+
+    response: bool
+              Response given by the user
+    """
+
+    while True:
+        reply = input(question+' (y/n): ').lower().strip()
+        if reply[:1] == 'y':
+            return True
+        if reply[:1] == 'n':
+            return False
+
+
+def add_labels(contrast, labels, output_file=ALL_CONTRASTS):
+    """
+    Adds all the passed labels to the selected contrast
+
+    Paramenters
+    -----------
+
+    contrast: str
+              Name of the contrast that will get the labels
+
+    labels: list of str
+            Labels that the user wants to add. The labels must exist as
+            columns in the file
+
+    output_file: str or path object
+                 Path to csv file where the new label database is to be saved
+                 with the changed
+    """
+    df = pd.read_csv(ALL_CONTRASTS, sep='\t')
+    con_index = df[df['contrast'] == contrast].index
+    for label in labels:
+        if label in df.columns:
+            df.at[con_index, label] = 1.0
+        else:
+            print(f"There is no label with the name {label}")
+            response = _binary_question(f"Do you want to add {label}? [y/n]: ")
```

Comment:
No, never render things uselessly interactive. This codebase is not designed for that.

Reply:
Done